File: backend/services/stigmer-server/pkg/domain/schedule/temporal/tick_workflow_test.go
package temporal

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"
	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/testsuite"
	"go.temporal.io/sdk/workflow"
)

// The tick workflow's contract, pinned against the cloud twin: the
// revalidation short-circuit, the verdict matrix, the two tracking
// bounds, and the backoff shape. workflow.Sleep is time-skipped by the
// test environment, so even the bounded-cycles test runs instantly.

const testScheduleID = "sch_01TESTTICK"

// Stub signatures matching the production activities (context first, as
// registered implementations receive it).
func stubRecordTick(_ context.Context, _ string, _ string) (TickOutcome, error) {
	return TickFired, nil
}
func stubStartScheduledRun(_ context.Context, _ string, _ string) (*RunStart, error) {
	return nil, nil
}
func stubPollExecutionPhase(_ context.Context, _ string) (RunPhase, error) {
	return PhaseRunning, nil
}
func stubRecordSuccessfulRun(_ context.Context, _ string) error { return nil }
func stubRecordFailedRun(_ context.Context, _ string, _ string, _ FailureKind) (*FailureRecorded, error) {
	return &FailureRecorded{}, nil
}

func newTickEnv(t *testing.T) *testsuite.TestWorkflowEnvironment {
	t.Helper()
	s := testsuite.WorkflowTestSuite{}
	env := s.NewTestWorkflowEnvironment()
	env.RegisterWorkflowWithOptions((&TickWorkflow{}).Run,
		workflow.RegisterOptions{Name: TickWorkflowType})
	env.RegisterActivityWithOptions(stubRecordTick,
		activity.RegisterOptions{Name: RecordTickActivityName})
	env.RegisterActivityWithOptions(stubStartScheduledRun,
		activity.RegisterOptions{Name: StartScheduledRunActivityName})
	env.RegisterActivityWithOptions(stubPollExecutionPhase,
		activity.RegisterOptions{Name: PollExecutionPhaseActivityName})
	env.RegisterActivityWithOptions(stubRecordSuccessfulRun,
		activity.RegisterOptions{Name: RecordSuccessfulRunActivityName})
	env.RegisterActivityWithOptions(stubRecordFailedRun,
		activity.RegisterOptions{Name: RecordFailedRunActivityName})
	return env
}

func started(executionID string, budgetMinutes int) *RunStart {
	return &RunStart{Outcome: RunStarted, ExecutionID: executionID,
		TrackingTimeoutMinutes: budgetMinutes}
}

func TestTick_SkippedTickNeverStartsARun(t *testing.T) {
	for _, outcome := range []TickOutcome{TickSkippedDeleted, TickSkippedDisabled, TickSkippedAutoPaused} {
		t.Run(string(outcome), func(t *testing.T) {
			env := newTickEnv(t)
			env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
				Return(outcome, nil)
			// No expectation on StartScheduledRun: an unexpected call fails the test.

			env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

			require.True(t, env.IsWorkflowCompleted())
			require.NoError(t, env.GetWorkflowError(),
				"a skipped tick is a successful no-op — the revalidation is what makes orphaned artifacts harmless")
			env.AssertNotCalled(t, StartScheduledRunActivityName, mock.Anything, mock.Anything, mock.Anything)
		})
	}
}

func TestTick_FiredTickPassesTheSameNominalTimeToTheRunStart(t *testing.T) {
	// The nominal time is the fire's identity: last_fire_at records it
	// and the execution name derives from it — the two idempotency keys
	// must never disagree, so both activities must receive one value.
	env := newTickEnv(t)
	var recordedAt, startedAt string
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(func(_ context.Context, _ string, nominal string) (TickOutcome, error) {
			recordedAt = nominal
			return TickFired, nil
		})
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(func(_ context.Context, _ string, nominal string) (*RunStart, error) {
			startedAt = nominal
			return &RunStart{Outcome: RunSkipped, TrackingTimeoutMinutes: 60}, nil
		})

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	require.NotEmpty(t, recordedAt)
	require.Equal(t, recordedAt, startedAt,
		"recordTick and startScheduledRun must see the SAME nominal fire time")
	parsed, err := time.Parse(time.RFC3339, recordedAt)
	require.NoError(t, err, "the nominal time crosses the activity boundary as RFC-3339")
	require.Zero(t, parsed.Nanosecond(), "nominal times are whole seconds — the workflow-id suffix granularity")
}

func TestTick_NominalTimeFromWorkflowIDSuffix(t *testing.T) {
	// Tier 2 of the derivation: Temporal appends the nominal time to the
	// artifact's base workflow id per fire. (Tier 1, the search
	// attribute, is not settable in the test framework — same limitation
	// the cloud twin records.)
	env := newTickEnv(t)
	env.SetStartWorkflowOptions(client.StartWorkflowOptions{
		ID: ArtifactID(testScheduleID) + "-2026-08-03T09:00:00Z",
	})
	var recordedAt string
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(func(_ context.Context, _ string, nominal string) (TickOutcome, error) {
			recordedAt = nominal
			return TickSkippedDisabled, nil
		})

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.Equal(t, "2026-08-03T09:00:00Z", recordedAt,
		"the workflow-id suffix IS the nominal fire time")
}

func TestTick_CompletedRunResetsTheStreak(t *testing.T) {
	env := newTickEnv(t)
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(TickFired, nil)
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(started("aex_01RUN", 60), nil)
	env.OnActivity(PollExecutionPhaseActivityName, mock.Anything, "aex_01RUN").
		Return(PhaseCompleted, nil)
	env.OnActivity(RecordSuccessfulRunActivityName, mock.Anything, testScheduleID).
		Return(nil)

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	env.AssertCalled(t, RecordSuccessfulRunActivityName, mock.Anything, testScheduleID)
	env.AssertNotCalled(t, RecordFailedRunActivityName,
		mock.Anything, mock.Anything, mock.Anything, mock.Anything)
}

func TestTick_TerminalFailuresFeedTheStreakWithThePhaseNamed(t *testing.T) {
	cases := []struct {
		phase  RunPhase
		reason string
	}{
		{PhaseFailed, "run aex_01RUN ended failed"},
		{PhaseCancelled, "run aex_01RUN ended cancelled"},
		{PhaseTerminated, "run aex_01RUN ended terminated"},
	}
	for _, tc := range cases {
		t.Run(string(tc.phase), func(t *testing.T) {
			env := newTickEnv(t)
			env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
				Return(TickFired, nil)
			env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
				Return(started("aex_01RUN", 60), nil)
			env.OnActivity(PollExecutionPhaseActivityName, mock.Anything, "aex_01RUN").
				Return(tc.phase, nil)
			var recordedReason string
			var recordedKind FailureKind
			env.OnActivity(RecordFailedRunActivityName, mock.Anything, testScheduleID, mock.Anything, mock.Anything).
				Return(func(_ context.Context, _ string, reason string, kind FailureKind) (*FailureRecorded, error) {
					recordedReason = reason
					recordedKind = kind
					return &FailureRecorded{ConsecutiveFailures: 1}, nil
				})

			env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

			require.True(t, env.IsWorkflowCompleted())
			require.NoError(t, env.GetWorkflowError())
			require.Equal(t, tc.reason, recordedReason, "the verdict names the terminal phase")
			require.Equal(t, FailureRunFailed, recordedKind)
		})
	}
}

func TestTick_BudgetExhaustionRecordsATimeoutWithoutCancellingTheRun(t *testing.T) {
	env := newTickEnv(t)
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(TickFired, nil)
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(started("aex_01HUNG", 1), nil) // one-minute budget
	env.OnActivity(PollExecutionPhaseActivityName, mock.Anything, "aex_01HUNG").
		Return(PhaseRunning, nil) // never terminal
	var recordedKind FailureKind
	env.OnActivity(RecordFailedRunActivityName, mock.Anything, testScheduleID, mock.Anything, mock.Anything).
		Return(func(_ context.Context, _ string, reason string, kind FailureKind) (*FailureRecorded, error) {
			recordedKind = kind
			require.Equal(t, "run aex_01HUNG did not finish within 1 minutes", reason)
			return &FailureRecorded{ConsecutiveFailures: 1}, nil
		})

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	require.Equal(t, FailureRunTimedOut, recordedKind)
}

func TestTick_TrackingIsBoundedByMaxCycles(t *testing.T) {
	// A misconfigured budget must not outlive the code-level backstop:
	// with an absurd budget and a never-terminal run, the loop exits
	// after at most MaxTrackingCycles polls and records a timeout —
	// the MaxRecoveryCycles convention: one test proves the loop works,
	// this one proves it terminates.
	env := newTickEnv(t)
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(TickFired, nil)
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(started("aex_01FOREVER", 1_000_000), nil)
	pollCount := 0
	env.OnActivity(PollExecutionPhaseActivityName, mock.Anything, "aex_01FOREVER").
		Return(func(_ context.Context, _ string) (RunPhase, error) {
			pollCount++
			return PhaseRunning, nil
		})
	env.OnActivity(RecordFailedRunActivityName, mock.Anything, testScheduleID, mock.Anything, mock.Anything).
		Return(&FailureRecorded{ConsecutiveFailures: 1}, nil)

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	require.Equal(t, MaxTrackingCycles, pollCount,
		"the loop polls exactly MaxTrackingCycles times before giving up")
}

func TestTick_GoneRunRecordsNoVerdict(t *testing.T) {
	// Deleting a run must not brick its schedule: no reset, no strike.
	env := newTickEnv(t)
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(TickFired, nil)
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(started("aex_01DELETED", 60), nil)
	env.OnActivity(PollExecutionPhaseActivityName, mock.Anything, "aex_01DELETED").
		Return(PhaseGone, nil)

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	env.AssertNotCalled(t, RecordSuccessfulRunActivityName, mock.Anything, mock.Anything)
	env.AssertNotCalled(t, RecordFailedRunActivityName,
		mock.Anything, mock.Anything, mock.Anything, mock.Anything)
}

func TestTick_StartFailuresFeedTheStreakWithoutTracking(t *testing.T) {
	cases := []struct {
		name    string
		outcome RunOutcome
		reason  string
	}{
		{"target missing", RunTargetMissing, "target agent acme/fee-bot not found"},
		{"refused", RunRefused, "run refused: insufficient credits"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			env := newTickEnv(t)
			env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
				Return(TickFired, nil)
			env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
				Return(&RunStart{Outcome: tc.outcome, TrackingTimeoutMinutes: 60,
					FailureReason: tc.reason}, nil)
			var recordedReason string
			var recordedKind FailureKind
			env.OnActivity(RecordFailedRunActivityName, mock.Anything, testScheduleID, mock.Anything, mock.Anything).
				Return(func(_ context.Context, _ string, reason string, kind FailureKind) (*FailureRecorded, error) {
					recordedReason = reason
					recordedKind = kind
					return &FailureRecorded{ConsecutiveFailures: 1}, nil
				})

			env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

			require.True(t, env.IsWorkflowCompleted())
			require.NoError(t, env.GetWorkflowError())
			require.Equal(t, tc.reason, recordedReason,
				"the start-failure copy flows into the streak verbatim")
			require.Equal(t, FailureStartFailed, recordedKind)
			env.AssertNotCalled(t, PollExecutionPhaseActivityName, mock.Anything, mock.Anything)
		})
	}
}

func TestTick_SkippedRunStartEndsTheTick(t *testing.T) {
	env := newTickEnv(t)
	env.OnActivity(RecordTickActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(TickFired, nil)
	env.OnActivity(StartScheduledRunActivityName, mock.Anything, testScheduleID, mock.Anything).
		Return(&RunStart{Outcome: RunSkipped, TrackingTimeoutMinutes: 60}, nil)

	env.ExecuteWorkflow(TickWorkflowType, testScheduleID)

	require.True(t, env.IsWorkflowCompleted())
	require.NoError(t, env.GetWorkflowError())
	env.AssertNotCalled(t, PollExecutionPhaseActivityName, mock.Anything, mock.Anything)
	env.AssertNotCalled(t, RecordFailedRunActivityName,
		mock.Anything, mock.Anything, mock.Anything, mock.Anything)
}

func TestTrackingBackoff_LinearTimesFiveCappedAtSixty(t *testing.T) {
	// recoveryBackoff's exact shape — the only backoff curve the
	// platform runs in production workflow code. Cap reached by cycle
	// twelve, bounding poll volume on long runs.
	require.Equal(t, 5*time.Second, trackingBackoff(1))
	require.Equal(t, 25*time.Second, trackingBackoff(5))
	require.Equal(t, 60*time.Second, trackingBackoff(12))
	require.Equal(t, 60*time.Second, trackingBackoff(240))
}

File: backend/services/stigmer-server-ts/src/temporal/schedule/__tests__/tick-workflow.test.ts
/**
 * Tick-workflow orchestration tests — port the scenario matrix of
 * pkg/domain/schedule/temporal/tick_workflow_test.go through a
 * TestWorkflowEnvironment with scripted activities.
 *
 * What these pin (the clock's orchestration contract):
 *   - skip outcomes end the tick without a run start;
 *   - start failures (TARGET_MISSING/REFUSED) record ONE failure with the
 *     START_FAILED kind and the deterministic reason;
 *   - a tracked run polls to COMPLETED → the success reset (never the
 *     failure recorder);
 *   - FAILED/CANCELLED/TERMINATED record the "run X ended <phase>" verdict;
 *   - GONE yields NO verdict (a deleted run must not brick its schedule);
 *   - budget exhaustion records RUN_TIMED_OUT with the byte-pinned reason
 *     and does NOT cancel the run;
 *   - a poll failure past retries fails the tick with NO verdict;
 *   - the 3-tier nominal-time derivation (workflow-id suffix, then
 *     workflow time truncated to whole seconds).
 *
 * Follows the invoke-workflow precedent: TestWorkflowEnvironment
 * .createLocal (needs the `temporal` CLI on PATH); every test skips
 * VISIBLY when the local test server cannot start — never a vacuous green
 * (the #18 panel lesson).
 */
import { afterAll, afterEach, beforeAll, describe, expect, it } from "vitest";

import {
  FAILURE_RUN_FAILED,
  FAILURE_RUN_TIMED_OUT,
  FAILURE_START_FAILED,
  PHASE_COMPLETED,
  PHASE_FAILED,
  PHASE_GONE,
  PHASE_RUNNING,
  RUN_ALREADY_STARTED,
  RUN_REFUSED,
  RUN_SKIPPED,
  RUN_STARTED,
  RUN_TARGET_MISSING,
  TICK_FIRED,
  TICK_SKIPPED_DISABLED,
  TICK_WORKFLOW_TYPE,
  type FailureKind,
  type RunPhase,
  type RunStart,
  type ScheduleTickActivities,
  type TickOutcome,
} from "../names.js";
import { MAX_TRACKING_CYCLES, trackingBackoffMs } from "../workflows/tick.js";

const TASK_QUEUE = "tick-workflow-test";
const WORKFLOWS_PATH = new URL("../workflows/index.ts", import.meta.url).pathname;

type TestWorkflowEnvironment = import("@temporalio/testing").TestWorkflowEnvironment;
type Worker = import("@temporalio/worker").Worker;

let env: TestWorkflowEnvironment | null = null;
let worker: Worker | null = null;
let workerRunPromise: Promise<void> | null = null;
let envReady = false;

// ─── Scriptable activity doubles ────────────────────────────────────────

interface FailureCall {
  scheduleId: string;
  reason: string;
  kind: FailureKind;
}

interface TickScript {
  tickOutcome: TickOutcome;
  /** Thrown from record-tick when set (nonRetryable — fails the tick fast). */
  tickError?: string;
  recordTickArgs: Array<{ scheduleId: string; nominal: string }>;
  runStart: RunStart;
  /** Consumed per poll, in order; the last one sticks. */
  pollResults: RunPhase[];
  /** Thrown from poll when set. */
  pollError?: string;
  pollCount: number;
  successCalls: string[];
  failureCalls: FailureCall[];
}

let script: TickScript;

function resetScript(): void {
  script = {
    tickOutcome: TICK_FIRED,
    recordTickArgs: [],
    runStart: {
      outcome: RUN_STARTED,
      executionId: "aex_tick",
      trackingTimeoutMinutes: 60,
      failureReason: "",
    },
    pollResults: [PHASE_COMPLETED],
    pollCount: 0,
    successCalls: [],
    failureCalls: [],
  };
}
resetScript();

async function nonRetryable(message: string): Promise<never> {
  const { ApplicationFailure } = await import("@temporalio/common");
  throw ApplicationFailure.nonRetryable(message);
}

// Typed as the REAL activity surface: a signature change in
// ScheduleTickActivities flags these doubles at compile time instead of
// silently decoupling the tests from the contract they pin.
function scriptedActivities(): ScheduleTickActivities {
  return {
    "stigmer/schedule/record-tick": async (scheduleId, nominal) => {
      script.recordTickArgs.push({ scheduleId, nominal });
      if (script.tickError !== undefined) {
        await nonRetryable(script.tickError);
      }
      return script.tickOutcome;
    },
    "stigmer/schedule/start-run": async () => script.runStart,
    "stigmer/schedule/poll-phase": async () => {
      if (script.pollError !== undefined) {
        await nonRetryable(script.pollError);
      }
      script.pollCount++;
      return script.pollResults.length > 1
        ? script.pollResults.shift()!
        : script.pollResults[0]!;
    },
    "stigmer/schedule/record-success": async (scheduleId) => {
      script.successCalls.push(scheduleId);
    },
    "stigmer/schedule/record-failure": async (scheduleId, reason, kind) => {
      script.failureCalls.push({ scheduleId, reason, kind });
      return { consecutiveFailures: script.failureCalls.length, paused: false };
    },
  };
}

let workflowSeq = 0;

async function runTick(options?: { workflowId?: string }): Promise<void> {
  if (!env) throw new Error("TestWorkflowEnvironment not initialized");
  workflowSeq++;
  const handle = await env.client.workflow.start(TICK_WORKFLOW_TYPE, {
    taskQueue: TASK_QUEUE,
    workflowId: options?.workflowId ?? `tick-test-${workflowSeq}-${Date.now()}`,
    args: ["sch_tick"],
  });
  await handle.result();
}

describe("schedule/tick workflow (TestWorkflowEnvironment)", () => {
  beforeAll(async () => {
    try {
      const { TestWorkflowEnvironment: TWE } = await import("@temporalio/testing");
      const { Worker: W } = await import("@temporalio/worker");
      env = await TWE.createLocal();
      worker = await W.create({
        connection: env.nativeConnection,
        taskQueue: TASK_QUEUE,
        workflowsPath: WORKFLOWS_PATH,
        activities: scriptedActivities(),
      });
      workerRunPromise = worker.run();
      envReady = true;
    } catch (error) {
      console.warn(
        `Temporal test server unavailable (tests will be skipped): ${error instanceof Error ? error.message : String(error)}`,
      );
      envReady = false;
    }
  }, 120_000);

  afterAll(async () => {
    if (worker) {
      worker.shutdown();
      await workerRunPromise?.catch(() => {});
    }
    if (env) await env.teardown();
  }, 30_000);

  afterEach(() => {
    resetScript();
  });

  it("a skipped tick ends without a run start or verdict", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.tickOutcome = TICK_SKIPPED_DISABLED;

    await runTick();
    expect(script.recordTickArgs).toHaveLength(1);
    expect(script.pollCount).toBe(0);
    expect(script.successCalls).toEqual([]);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it.for([
    [RUN_TARGET_MISSING, "target agent acme/gone not found"],
    [RUN_REFUSED, "run refused: gate said no"],
  ] as const)(
    "a %s start records ONE START_FAILED verdict with the deterministic reason",
    { timeout: 30_000 },
    async ([outcome, reason], testCtx) => {
      if (!envReady) return testCtx.skip();
      script.runStart = {
        outcome,
        executionId: "",
        trackingTimeoutMinutes: 60,
        failureReason: reason,
      };

      await runTick();
      expect(script.failureCalls).toEqual([
        { scheduleId: "sch_tick", reason, kind: FAILURE_START_FAILED },
      ]);
      expect(script.successCalls).toEqual([]);
      expect(script.pollCount).toBe(0);
    },
  );

  it("a SKIPPED run start ends the tick quietly", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.runStart = {
      outcome: RUN_SKIPPED,
      executionId: "",
      trackingTimeoutMinutes: 60,
      failureReason: "",
    };
    await runTick();
    expect(script.failureCalls).toEqual([]);
    expect(script.pollCount).toBe(0);
  }, 30_000);

  it("tracks a started run to COMPLETED and resets the streak", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.pollResults = [PHASE_COMPLETED];

    await runTick();
    expect(script.pollCount).toBe(1);
    expect(script.successCalls).toEqual(["sch_tick"]);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it("re-polls a still-RUNNING run through the backoff sleep to its verdict", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    // Two polls with one real cycle-1 backoff (5s) between them — the
    // sleep→re-poll path the single-poll scenarios never execute.
    script.pollResults = [PHASE_RUNNING, PHASE_COMPLETED];

    await runTick();
    expect(script.pollCount).toBe(2);
    expect(script.successCalls).toEqual(["sch_tick"]);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it("an ALREADY_STARTED retry tracks exactly like STARTED", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.runStart = {
      outcome: RUN_ALREADY_STARTED,
      executionId: "aex_tick",
      trackingTimeoutMinutes: 60,
      failureReason: "",
    };
    await runTick();
    expect(script.successCalls).toEqual(["sch_tick"]);
  }, 30_000);

  it("records the 'run X ended failed' verdict on a FAILED phase", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.pollResults = [PHASE_FAILED];

    await runTick();
    expect(script.failureCalls).toEqual([
      {
        scheduleId: "sch_tick",
        reason: "run aex_tick ended failed",
        kind: FAILURE_RUN_FAILED,
      },
    ]);
    expect(script.successCalls).toEqual([]);
  }, 30_000);

  it("GONE yields no verdict (a deleted run must not brick its schedule)", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.pollResults = [PHASE_GONE];

    await runTick();
    expect(script.successCalls).toEqual([]);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it("budget exhaustion records RUN_TIMED_OUT with the byte-pinned reason (run NOT cancelled)", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    // A zero-minute budget: the deadline passes after the first RUNNING
    // poll (the poll-first loop never sleeps past its own deadline).
    script.runStart = {
      outcome: RUN_STARTED,
      executionId: "aex_tick",
      trackingTimeoutMinutes: 0,
      failureReason: "",
    };
    script.pollResults = [PHASE_RUNNING];

    await runTick();
    expect(script.pollCount).toBe(1);
    expect(script.failureCalls).toEqual([
      {
        scheduleId: "sch_tick",
        reason: "run aex_tick did not finish within 0 minutes",
        kind: FAILURE_RUN_TIMED_OUT,
      },
    ]);
  }, 30_000);

  it("a poll failure past retries fails the tick with NO verdict", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.pollError = "db unreachable";

    await expect(runTick()).rejects.toThrow();
    expect(script.successCalls).toEqual([]);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it("a record-tick failure fails the tick before any run start", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.tickError = "row unreadable";

    await expect(runTick()).rejects.toThrow();
    expect(script.pollCount).toBe(0);
    expect(script.failureCalls).toEqual([]);
  }, 30_000);

  it("derives the nominal fire time from the workflow-id suffix (tier 2)", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.tickOutcome = TICK_SKIPPED_DISABLED;

    await runTick({ workflowId: "schedule/tick/sch_tick-2026-08-25T09:30:00Z" });
    expect(script.recordTickArgs[0]?.nominal).toBe("2026-08-25T09:30:00Z");
  }, 30_000);

  it("falls back to workflow time truncated to whole seconds (tier 3)", async (testCtx) => {
    if (!envReady) return testCtx.skip();
    script.tickOutcome = TICK_SKIPPED_DISABLED;

    await runTick();
    const nominal = script.recordTickArgs[0]?.nominal ?? "";
    // RFC-3339, whole seconds, UTC — the fire-identity format.
    expect(nominal).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/);
  }, 30_000);
});

describe("tick loop bounds (plain unit)", () => {
  it("pins MAX_TRACKING_CYCLES at 240 (the recorded-history backstop)", () => {
    expect(MAX_TRACKING_CYCLES).toBe(240);
  });

  it("pins the tracking backoff curve: linear cycle×5s, capped at 60s by cycle twelve", () => {
    expect(trackingBackoffMs(1)).toBe(5_000);
    expect(trackingBackoffMs(2)).toBe(10_000);
    expect(trackingBackoffMs(11)).toBe(55_000);
    expect(trackingBackoffMs(12)).toBe(60_000);
    expect(trackingBackoffMs(13)).toBe(60_000);
    expect(trackingBackoffMs(240)).toBe(60_000);
  });
});
